app/models/sample.rb

```diff
@@ -328,6 +337,62 @@ def initiate_s3_cp
     LogUtil.log_err_and_airbrake("Failed to upload S3 sample '#{name}' (#{id}): #{e}")
   end
 
+  def initiate_basespace_file_transfer
+    # Only continue if the sample status is CREATED and not yet UPLOADED.
+    return unless status == STATUS_CREATED
+
+    Rails.logger.info("HERE #{basespace_dataset_id} #{basespace_access_token}")
+
+    files = files_for_basespace_dataset(basespace_dataset_id, basespace_access_token)
+
+    raise "Could not fetch Basespace files for dataset #{basespace_dataset_id} for sample '#{name}' (#{id})" if files.blank?
+
+    files.each do |file|
+      max_tries = 3
+      try = 0
+      time_between_tries = 10
+
+      Rails.logger.info("Starting upload of sample '#{name}' (#{id}) file '#{file[:name]}' from Basespace")
+      while try < max_tries
+        success = download_from_basespace_to_s3(file[:download_path], sample_input_s3_path, file[:name])
+
+        if success
+          break
+        else
+          Rails.logger.error("Try ##{try}: Upload of sample '#{name}' (#{id}) file '#{file[:name]}' from Basespace failed")
+          try += 1
+          if try < max_tries
+            Kernel.sleep(time_between_tries)
+          end
+        end
+      end
+
+      raise "Upload of sample '#{name}' (#{id}) file '#{file[:name]} from Basespace failed after #{max_tries} retries" unless try < max_tries
+
+      input_file = InputFile.new(
+        name: file[:name],
+        source_type: InputFile::SOURCE_TYPE_BASESPACE,
+        source: file[:source_path]
+      )
+
+      input_files << input_file
+    end
+
+    self.status = STATUS_UPLOADED
+    save
+  rescue => e
+    LogUtil.log_err_and_airbrake("SampleUploadFailedEvent: #{e}")
+
+    pr = new_failed_pipeline_run(self)
```

Comment:
It's not totally clear to me why we need to create a failed pipeline object when there is a failed basepace transfer. Is it because the transfer happens asynchronously after the user clicks submit on their samples? And so the pipeline object is the most convenient channel for reporting an error? And if so, why don't we do the same for s3 transfers that also appear to happen after the user clicks submit?

Reply:
That's a good point. I did this to be consistent with how we do our error handling elsewhere. In all other cases besides file upload, the failure would happen during the pipeline.

But I could see an argument for adding an `upload_error` field to the sample, since this technically happens independent of any pipeline.

Reply:
I ended up using an `upload_error` field on the sample, which I think makes the most sense.